CSP/rules.py

```diff
@@ -0,0 +1,92 @@
+from find_domain import expand_harmony
+
+
+def fulfills_harmony(voices, harmony, problem, names):
+    for i in range(0, len(harmony)):
+        required_pitches = expand_harmony(harmony[i])
+        variables = []
+        for j in range(0, 4):
+            pitch = voices[j][i]
+            if pitch == '??':
+                variables.append(names[j] + str(i + 1))  # Keep track of which voices need to be filled
+            else:
+                try:
+                    required_pitches.remove(pitch % 12)  # Keep track of pitches that are already present
+                except ValueError:
+                    pass
+
+        # Open to suggestions on ways to make this cleaner instead of the multiple elif statements
+        if len(variables) == 1:
+            problem.addConstraint(lambda x: (x % 12) in required_pitches, [variables[0]])
+        elif len(variables) == 2:
+            problem.addConstraint(lambda x, y, r=required_pitches: set(r).issubset({x % 12, y % 12}),
+                                  [variables[0], variables[1]])
+        elif len(variables) == 3:
+            problem.addConstraint(lambda x, y, z, r=required_pitches: set(r).issubset({x % 12, y % 12, z % 12}),
+                                  [variables[0], variables[1], variables[2]])
+        elif len(variables) == 4:
+            problem.addConstraint(lambda w, x, y, z, r=required_pitches: set(r).issubset({w % 12, x % 12, y % 12, z % 12}),
+                                  [variables[0], variables[1], variables[2], variables[3]])
+
+
+def parallels(voices, problem, names, interval):
+    for i in range(0, len(voices[0]) - 1):  # Doesn't apply to last pitches (final cadence)
```

Comment:
Oops, missed this in the validator.  Making sure you want the validator to stop before last pitches too?